### ISSUE TITLE
Please consider leaving closing tags untouched, or document the formatter behavior

### ISSUE THREAD

**Issue by oliviercailloux** [google/google-java-format#647]
I would like to argue against the formatter removing my closing HTML tags.

I have no problem with the formatter changing my code as required to make it consistent to Google Java Style Guide, as this is its declared aim: [to reformat Java source code “to comply with Google Java Style”](https://github.com/google/google-java-format/blob/master/README.md#google-java-format).

However, it is reasonable to expect that it will not significantly change my code beyond what is required to make it consistent to Google Java Style Guide (and obviously reasonable generalizations of it). 

As the formatter is declared to implement a given normative reference (the Google Java Style Guide), I think it would be clearer to refrain from adopting more opinionated decisions inside the formatter itself. These decisions should be taken at the level of the Google Java Style Guide.

Desiring to omit closing HTML tags is, I suppose, a reasonable decision in some contexts, but desiring to have them is reasonable as well, and a very common behavior among web developers (as illustrated in several [StackOverflow discussions](https://stackoverflow.com/q/7125354) and in #61). Therefore, they should be left untouched.

If the designers want to make the formatter opinionated about some issue beyond what the normative reference prescribes, such as the closing tags, it should at least be clearly documented.

In issue #7, cpovirk wrote: “We presumably don't want to remove optional closing tags (`</p>`, `</li>`), but we could.” I am not sure why this suggestion was not followed, or even whether it is a conscious collegial choice.

**Comment by tbroyer** [comment]
Somewhat related: https://github.com/google/google-java-format/wiki/FAQ#why-didnt-it-leave-my-perfectly-valid-code-alone

**Comment by oliviercailloux** [comment]
> Somewhat related: https://github.com/google/google-java-format/wiki/FAQ#why-didnt-it-leave-my-perfectly-valid-code-alone

… though in my view I am not talking about a formatting choice. A closing tag is code (HTML code), not formatting. Removing it is a decision that changes my code.

I agree that making the code consistent with the Google Java Style does sometimes require to change the code itself (such as [adding the opening `<p>` tags](https://google.github.io/styleguide/javaguide.html#s7.1.2-javadoc-paragraphs)), but such decisions should be limited to what is explicitly documented, and are of a different nature than deciding to “forget” most of the existing formatting (a perfectly reasonable decision that I am not arguing against, unless considering that HTML code is part of the formatting).

**Comment by kevinb9n** [comment]
So far, there's been more assertion in this report than argument, and the arguments have been predicated on a different set of values than the ones we hold and which have (more than not) led to the tool being the way it is.

I think it will help if you set all the "you-should" aside and focus on the _why_ behind it. And I think you can go one of two main routes:

* why it makes sense given the same values that the formatter already lives by, not by the _different_ values that you want it to
* why the values themselves should be changed

It's not always easy to *start out* this way, because you usually don't know the values in full detail. We have made I think a reasonable attempt to make them clear in the faq, but we didn't go into this particular case:

(An example of a value decision.) We have to sort out of which of these matters more to us:

* All else being equal, we'd like documentation to be as readable as possible in source form (more or less, the closer it looks to markdown the better)
* All else being equal, we'd like to generate valid HTML, or at least not stop users from being able to do so

(I do agree that both of these are positives.)

I have *so far* never valued the second one very highly, for one reason: the Javadoc html could be deterministically post-processed to add those closing tags (a &lt;p> or &lt;li> never legitimately directly contains another like itself). In fact, to the extent that the issues brought up in the SO post are real and impactful, I would kind of expect that a javadoc.exe feature request is where we should clamor for this behavior. (Someone might be willing to find if such a FR ever existed.) I also expect that we'd be seeing this phrased as actual observed bugs rather than two GJF issues that are both written from the "you just should" perspective.

Obviously, part of this is that I subjectively do find the closing tags in javadoc to be significantly obtrusive, and you may well not. We would have to probably feel like a survey is worthwhile if we wanted to do something here. But I mostly feel like there is a best-of-both-worlds solution through tools anyway. Open to your further thoughts.

**Comment by tbroyer** [comment]
Fwiw, GJF currently focuses on whitespace changes (see https://github.com/google/google-java-format/issues/243#issuecomment-356315049), and the behavior in Javadoc is thus different from what can be seen in code (those closing tags could be passed through untouched, but are willfully removed).
I for one, am OK with removing closing tags in Javadoc that can safely be omitted, and would be pleased to see more non-whitespace changes in code (e.g. add braces)
Speaking of HTML, if one thinks about it in terms of the Infoset (elements and text nodes), then tags *are* formatting; a detail of the way one can serialize that Infoset. It appears that GJF does not actually *see* the Java code that way though (otherwise it would probably add those braces). In a few words, looking at what GJF currently does to draw conclusions about what it should do in other situations is not as easy as it seems.

**Comment by kevinb9n** [comment]
Of course. I had tried to make that clear (right after the first bullet list in my comment).
